homework-submission/react-hyf/react-hyf-client/src/App.js

```diff
@@ -0,0 +1,67 @@
+import React, { Component } from 'react';
+import './App.css';
+import Header from './components/Header';
+import TodoInput from './components/TodoInput';
+import TodoItem from './components/TodoItem';
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: [],
+      nextId: 5,
+    };
+
+    this.addTodo = this.addTodo.bind(this);
+    this.removeTodo = this.removeTodo.bind(this);
+  }
+
+  componentDidMount() {
+    this.fetchTodos().then(todos => {
+      this.setState({
+        todos,
+      });
+    });
+  }
+
+  fetchTodos = async () => {
+    const response = await fetch('/todos');
+    const data = await response.json();
+    return data;
+  };
+
+  addTodo(todoText) {
+    let todos = [...this.state.todos];
+    todos.push({ id: this.state.nextId, text: todoText });
```

Comment:
with this line you are mutating the state directly, you may use instead `.concat` or including it in the previous line as:
```javascript
// Option #1
let todos = [...this.state.todos, { id: this.state.nextId, text: todoText }];

// Option #2
let todos = this.state.todos.concat({ id: this.state.nextId, text: todoText });
```

- You may tell: But this is a shadow copy!

The answer is Well yes it's But it will work only if you have a one level array as `['hello', 567, 'this will be copied']` because this is a reference and you are copying the first layer.

While `[{name: "jawhar"}]` it won't work because it copied the reference that refers to the Object that includes `name: "jawhar"`.